src/io/iohk/scalanet/peergroup/TCPPeerGroup.scala

```diff
@@ -1,142 +1,224 @@
-    Codec[PDU[MessageType]]
-  }
-}
+  private class ClientChannelImpl[M](inetSocketAddress: InetSocketAddress, clientBootstrap: Bootstrap)(
+      implicit codec: Codec[M]
+  ) extends Channel[InetMultiAddress, M] {
+
+    private val log = LoggerFactory.getLogger(getClass)
+
+    val to: InetMultiAddress = InetMultiAddress(inetSocketAddress)
+
+    private val activation = Promise[ChannelHandlerContext]()
+    private val activationF = activation.future
+    private val deactivation = Promise[Unit]()
+    private val deactivationF = deactivation.future
+
+    private val messageSubject = ReplaySubject[M]()
+
+    private val bootstrap: Bootstrap = clientBootstrap
+      .clone()
+      .handler(new ChannelInitializer[SocketChannel]() {
+        def initChannel(ch: SocketChannel): Unit = {
+          ch.pipeline()
+            .addLast("frameEncoder", new LengthFieldPrepender(4))
+            .addLast("frameDecoder", new LengthFieldBasedFrameDecoder(Int.MaxValue, 0, 4, 0, 4))
+            .addLast(new ByteArrayEncoder())
+            .addLast(new ChannelInboundHandlerAdapter() {
+              override def channelActive(ctx: ChannelHandlerContext): Unit = {
+                log.debug(
+                  s"Creating client channel from ${ctx.channel().localAddress()} " +
+                    s"to ${ctx.channel().remoteAddress()} with channel id ${ctx.channel().id}"
+                )
+                activation.success(ctx)
+              }
+
+              override def channelInactive(ctx: ChannelHandlerContext): Unit = {
+                deactivation.success(())
+              }
+            })
+            .addLast(new MessageNotifier[M](messageSubject))
+        }
+      })
 
-object TCPPeerGroup {
+    def initialize: Task[ClientChannelImpl[M]] = toTask(bootstrap.connect(inetSocketAddress)).map(_ => this)
 
-  case class Config(bindAddress: InetSocketAddress, processAddress: InetSocketAddress)
+    override def sendMessage(message: M): Task[Unit] = {
 
-  object Config {
-    def apply(bindAddress: InetSocketAddress): Config = Config(bindAddress, bindAddress)
+      Task
+        .fromFuture(activationF)
+        .map(ctx => {
+          log.debug(
+            s"Processing outbound message from local address ${ctx.channel().localAddress()} " +
+              s"to remote address ${ctx.channel().remoteAddress()} via channel id ${ctx.channel().id()}"
+          )
+          ctx.writeAndFlush(Unpooled.wrappedBuffer(codec.encode(message)))
+        })
+        .map(_ => ())
+    }
+
+    override def in: Observable[M] = messageSubject
+
+    override def close(): Task[Unit] = {
+      messageSubject.onComplete()
+      Task
+        .fromFuture(activationF)
+        .flatMap(ctx => toTask(ctx.close()))
+        .flatMap(_ => Task.fromFuture(deactivationF))
+    }
   }
 
-  case class PDU[T](replyTo: InetSocketAddress, sdu: T)
+  private class MessageNotifier[M](val messageSubject: Subject[M, M])(implicit codec: Codec[M])
```

Comment:
Shouldn't user of of library have some control of error handling ? For example, currently when the exception will be thrown during handling of inbound Netty will log the fact that the exception wasn’t handled as method `exceptionCaught()` is not ovverriden in any `ChannelInboundHandlerAdapter`